File: cms-app-crawler-ts/src/base/BasePage.ts
import type { Page, Locator as PwLocator } from 'playwright';
import { Locator } from './Locator';
import { SelectorRegistry } from '../utils/SelectorRegistry';

export abstract class BasePage {
  protected readonly _page: Page;

  constructor(page: Page) {
    this._page = page;
  }

  async waitForReady(): Promise<void> {
    await this._page.waitForLoadState('domcontentloaded');
    await this._page.waitForLoadState('networkidle').catch(() => {});
  }

  async click(locator: Locator): Promise<void> {
    const el = await this._findElement(locator);
    await el.scrollIntoViewIfNeeded();
    try {
      await el.click();
    } catch {
      await this._page.evaluate((el) => (el as HTMLElement).click(), await el.elementHandle());
    }
  }

  async fill(locator: Locator, value: string): Promise<void> {
    const el = await this._findElement(locator);
    await el.fill('');
    await el.fill(value);
  }

  async selectOption(locator: Locator, value: string): Promise<void> {
    const el = await this._findElement(locator);
    await el.selectOption(value);
  }

  async assertVisible(locator: Locator): Promise<void> {
    const el = await this._findElement(locator);
    const visible = await el.isVisible();
    if (!visible) {
      throw new Error(`Expected element not visible: [${locator.getSelectorKey()}] on: ${await this._page.title()}`);
    }
  }

  async assertPageLoaded(): Promise<this> {
    await this.waitForReady();
    return this;
  }

  async assertTitle(expected: string): Promise<void> {
    const actual = await this._page.title();
    if (!actual.includes(expected)) {
      throw new Error(`Expected title: [${expected}] got: [${actual}]`);
    }
  }

  async getText(locator: Locator): Promise<string | null> {
    const el = await this._findElement(locator);
    return el.textContent();
  }

  async getCurrentUrl(): Promise<string> {
    return this._page.url();
  }

  protected async _findElement(locator: Locator): Promise<PwLocator> {
    const selectorKey = locator.getSelectorKey();
    const primary = locator.toPlaywrightSelector();
    const el = this._page.locator(primary).first();
    if (await el.count() > 0) {
      return el;
    }
    const fallbacks = SelectorRegistry.getFallbacks(selectorKey);
    for (const fb of fallbacks) {
      const fbSel = this._resolveSelector(fb);
      const fbEl = this._page.locator(fbSel).first();
      if (await fbEl.count() > 0) {
        console.warn(`Self-healing: [${selectorKey}] not found, used fallback [${fb}] on page: ${await this._page.title()}`);
        return fbEl;
      }
    }
    throw new Error(`Element not found: [${selectorKey}] and all ${fallbacks.length} fallback(s) failed. Page: ${await this._page.title()}`);
  }

  private _resolveSelector(sk: string): string {
    if (!sk) return '*';
    if (sk.startsWith('xpath='))       return sk;
    if (sk.startsWith('#'))            return sk;
    if (sk.startsWith("[name='"))      return sk;
    if (sk.startsWith('data-testid=')) return `[data-testid='${sk.substring(12)}']`;
    if (sk.startsWith('aria-label='))  return `[aria-label='${sk.substring(11)}']`;
    return sk;
  }
}

File: cms-app-crawler/src/main/java/base/BasePage.java
package base;

import org.openqa.selenium.*;
import org.openqa.selenium.support.ui.*;
import utils.SelectorRegistry;
import java.time.Duration;
import java.util.List;
import java.util.logging.Logger;

public abstract class BasePage {

    protected WebDriver driver;
    private static final Logger log = Logger.getLogger(BasePage.class.getName());

    public BasePage(WebDriver driver) {
        this.driver = driver;
        waitForReady();
    }

    protected void waitForReady() {
        new WebDriverWait(driver, Duration.ofSeconds(30))
            .until(d -> ((JavascriptExecutor) d)
                .executeScript("return document.readyState").equals("complete"));
    }

    protected void click(Locator locator) {
        WebElement el = findElement(locator);
        ((JavascriptExecutor) driver).executeScript("arguments[0].scrollIntoView({block:'center'});", el);
        try {
            el.click();
        } catch (ElementClickInterceptedException e) {
            ((JavascriptExecutor) driver).executeScript("arguments[0].click();", el);
        }
    }

    protected void fill(Locator locator, String value) {
        WebElement el = findElement(locator);
        el.clear();
        el.sendKeys(value);
    }

    protected void select(Locator locator, String value) {
        new Select(findElement(locator)).selectByVisibleText(value);
    }

    protected void assertVisible(Locator locator) {
        boolean visible = false;
        try { visible = findElement(locator).isDisplayed(); }
        catch (Exception ignored) {}
        if (!visible) throw new AssertionError(
            "Expected element not visible: [" + locator + "] on: " + driver.getTitle());
    }

    public BasePage assertPageLoaded() {
        waitForReady();
        return this;
    }

    protected void assertTitle(String expected) {
        String actual = driver.getTitle();
        if (!actual.contains(expected)) throw new AssertionError(
            "Expected title: [" + expected + "] got: [" + actual + "]");
    }

    protected String getText(Locator locator) {
        return findElement(locator).getText();
    }

    protected String getCurrentUrl() { return driver.getCurrentUrl(); }

    private WebElement findElement(Locator locator) {
        String selectorKey = locator.getSelectorKey();

        // Try the primary selector first
        try {
            return driver.findElement(resolveBy(locator));
        } catch (NoSuchElementException primary) {
            // Primary failed — try fallbacks from registry in priority order
            List<String> fallbacks = SelectorRegistry.getFallbacks(selectorKey);
            for (String fallback : fallbacks) {
                try {
                    WebElement el = driver.findElement(resolveByString(fallback));
                    log.warning(String.format(
                        "Self-healing: [%s] not found, used fallback [%s] on page: %s",
                        selectorKey, fallback, driver.getTitle()));
                    return el;
                } catch (NoSuchElementException ignored) {}
            }
            // All fallbacks exhausted
            throw new AssertionError(String.format(
                "Element not found: [%s] and all %d fallback(s) failed. Page: %s",
                selectorKey, fallbacks.size(), driver.getTitle()));
        }
    }

    private By resolveBy(Locator locator) {
        return resolveByString(locator.getSelectorKey());
    }

    // Resolves the selectorKey string format used by the registry
    private By resolveByString(String sk) {
        if (sk == null || sk.isEmpty()) return By.cssSelector("*");
        if (sk.startsWith("xpath="))             return By.xpath(sk.substring(6));
        if (sk.startsWith("#"))                  return By.id(sk.substring(1));
        if (sk.startsWith("[name='"))            return By.name(sk.replaceAll("\\[name='([^']+)'\\]", "$1"));
        if (sk.startsWith("data-testid="))       return By.cssSelector("[data-testid='" + sk.substring(12) + "']");
        if (sk.startsWith("aria-label="))        return By.cssSelector("[aria-label='" + sk.substring(11) + "']");
        if (sk.startsWith("input[type='"))       return By.cssSelector(sk);
        return By.cssSelector(sk);
    }
}
